File: Problem3.py
#T.C: O(N +M) where N is string and M is pattern
#S.C: O(1) 

class Solution:

    def wordPattern(self, pattern: str, s: str) -> bool:
        s = s.split(' ')
    
        if len(pattern) != len(s):
            return False

        dict_s = {}
        dict_p = {}

        for i in range(0, len(pattern)):
            pchar = pattern[i]
            schar = s[i]
        
            if pchar not in dict_p:
                if schar in dict_s:
                    return False
                else:
                    dict_p[pchar] = schar
                    dict_s[schar] = pchar
            else:
                if dict_p[pchar] != schar:
                    return False
        
        return True

File: problem-3.go
/*time complexity : O(n)
space complexity : O(3*n)*/
import "strings"

func wordPattern(pattern string, s string) bool {

	split := strings.Split(s, " ")
	if len(split) != len(pattern) {
		return false
	}
	patternM := make(map[byte]string)
	stringM := make(map[string]byte)
	for i := 0; i < len(split); i++ {
		pVal, pOkay := patternM[pattern[i]]
		sVal, sOkay := stringM[split[i]]
		if !pOkay && !sOkay {
			patternM[pattern[i]] = split[i]
			stringM[split[i]] = pattern[i]
			continue
		}
		if (pVal != split[i]) || sVal != pattern[i] {
			return false
		}
	}
	return true
}
